include/ginkgo/core/stop/residual_norm.hpp

```diff
@@ -77,17 +100,200 @@ class ResidualNorm
     {}
 
     explicit ResidualNorm(std::shared_ptr<const gko::Executor> exec,
-                          remove_complex<ValueType> tolerance)
+                          remove_complex<ValueType> reduction_factor)
         : EnablePolymorphicObject<ResidualNorm, Criterion>(exec),
           device_storage_{exec, 2},
-          tolerance_{tolerance}
+          reduction_factor_{reduction_factor}
     {}
 
+    explicit ResidualNorm(const Factory *factory, const CriterionArgs &args)
+        : EnablePolymorphicObject<ResidualNorm, Criterion>(
+              factory->get_executor()),
+          device_storage_{factory->get_executor(), 2},
+          parameters_{factory->get_parameters()},
+          reduction_factor_{factory->get_parameters().reduction_factor}
+    {
+        this->baseline_rhs_norm_ =
+            factory->get_parameters().baseline == mode::rhs_norm;
+        this->baseline_init_resnorm_ =
+            factory->get_parameters().baseline == mode::initial_resnorm;
+        this->baseline_absolute_ =
+            factory->get_parameters().baseline == mode::absolute;
```

Comment:
I don't understand why you still use these. Is it not easier to store the `enum` do a `switch` `case` over the options? This would also prevent the future error of having multiple of these `bool` set to `true`.